Support for `gast` with Python 3.12

### bug描述 Describe the Bug

PaddlePaddle utilizes `gast` for handling Python Abstract Syntax Trees (AST). With the release of Python 3.12, changes to Python’s grammar have introduced new syntax, such as `type_param`, which is now part of [Python's AST](https://docs.python.org/3/library/ast.html#function-and-class-definitions). The [gast](https://github.com/serge-sans-paille/gast/tree/0.6.0) has updated with this grammar. However, since PaddlePaddle does not use `gast` as a submodule, it has not kept pace with the latest updates in `gast`. 

The test `test_gast_with_compatibility` fails when executed in a Python 3.12 environment.

**Error Log:**
```
1382: ======================================================================
1382: FAIL: test_param_of_func (test_gast_with_compatibility.TestPythonCompatibility.test_param_of_func)
1382: Param in func is ast.Name in PY2, but ast.arg in PY3.
1382: ----------------------------------------------------------------------
1382: Traceback (most recent call last):
1382:   File "/home/fralin/workspace/paddle/paddle-develop/build/test/legacy_test/test_gast_with_compatibility.py", line 143, in test_param_of_func
1382:     self._check_compatibility(source, target)
1382:   File "/home/fralin/workspace/paddle/paddle-develop/build/test/legacy_test/test_gast_with_compatibility.py", line 128, in _check_compatibility
1382:     self.assertEqual(source_dump, target_dump)
1382: AssertionError: "Modu[244 chars]d())))], decorator_list=[])], type_ignores=[])" != "Modu[244 chars]d())))], decorator_list=[], type_params=[])], type_ignores=[])"
1382: Diff is 1187 characters long. Set self.maxDiff to None to see it.
1382:
1382: ----------------------------------------------------------------------
1382: Ran 5 tests in 0.003s
```

The error arises because the `type_params` syntax introduced in Python 3.12 is not recognized by the existing `gast` version in PaddlePaddle, causing mismatches in AST comparisons during testing.

Although updating `gast` to a newer version could potentially resolve the issue, there are additional code adjustments required within PaddlePaddle’s JIT modules. For example, the following locations need to be updated to handle grammar changes in Python 3.12:  [utils.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/jit/dy2static/utils.py#L309) [origin_info.py](https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/jit/dy2static/origin_info.py#L122).

### 其他补充信息 Additional Supplementary Information

_No response_

--- a/python/paddle/jit/dy2static/origin_info.py
+++ b/python/paddle/jit/dy2static/origin_info.py
@@ -104,7 +104,7 @@ class OriginInfoAttacher(gast.NodeTransformer):
     def visit(self, node):
         if isinstance(node, gast.FunctionDef):
             self.current_func.append(node)
-        if hasattr(node, "lineno"):
+        if getattr(node, "lineno", None) is not None:
             self._attach_origin_info(node)
         self.generic_visit(node)
 

--- a/python/paddle/jit/dy2static/transformers/decorator_transformer.py
+++ b/python/paddle/jit/dy2static/transformers/decorator_transformer.py
@@ -127,7 +127,7 @@ class DecoratorTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
-            type_param=[],
+            type_params=[],
         )
 
         args = [arg.id for arg in node.args.args]

--- a/python/paddle/jit/dy2static/transformers/decorator_transformer.py
+++ b/python/paddle/jit/dy2static/transformers/decorator_transformer.py
@@ -127,6 +127,7 @@ class DecoratorTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_param=[],
         )
 
         args = [arg.id for arg in node.args.args]

--- a/python/paddle/jit/dy2static/transformers/loop_transformer.py
+++ b/python/paddle/jit/dy2static/transformers/loop_transformer.py
@@ -587,6 +587,7 @@ class LoopTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_params=[],
         )
         new_stmts.append(condition_func_node)
 
@@ -607,6 +608,7 @@ class LoopTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_params=[],
         )
         new_stmts.append(body_func_node)
 
@@ -669,6 +671,7 @@ class LoopTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_params=[],
         )
 
         new_stmts.append(condition_func_node)
@@ -689,6 +692,7 @@ class LoopTransformer(BaseTransformer):
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_params=[],
         )
         new_stmts.append(body_func_node)
 

--- a/python/paddle/jit/dy2static/transformers/return_transformer.py
+++ b/python/paddle/jit/dy2static/transformers/return_transformer.py
@@ -251,6 +251,7 @@ class SingleReturnTransformer(BaseTransformer):
                     )
                 ],
                 value=gast.Constant(kind=None, value=None),
+                type_comment=None,
             )
             node.body.insert(0, assign_return_value_node)
 
@@ -368,6 +369,7 @@ class SingleReturnTransformer(BaseTransformer):
                         )
                     ],
                     value=return_node.value,
+                    type_comment=None,
                 )
             )
             return_origin_info = getattr(return_node, ORIGIN_INFO, None)

--- a/python/paddle/jit/dy2static/transformers/typehint_transformer.py
+++ b/python/paddle/jit/dy2static/transformers/typehint_transformer.py
@@ -45,5 +45,9 @@ class TypeHintTransformer(BaseTransformer):
     def visit_AnnAssign(self, node):
         if node.value is None:
             return None
-        assign_node = gast.Assign(targets=[node.target], value=node.value)
+        assign_node = gast.Assign(
+            targets=[node.target],
+            value=node.value,
+            type_comment=None,
+        )
         return assign_node

--- a/python/paddle/jit/dy2static/transformers/utils.py
+++ b/python/paddle/jit/dy2static/transformers/utils.py
@@ -128,6 +128,7 @@ def create_function_def_node(nodes, name, input_args, return_name_ids):
         decorator_list=[],
         returns=None,
         type_comment=None,
+        type_params=[],
     )
     return func_def_node
 
@@ -137,7 +138,11 @@ def create_assign_node(name, node):
     Creates a `gast.Assign` node by given name_id as target and node as value.
     """
     targets = generate_name_node(name, ctx=gast.Store())
-    assign_node = gast.Assign(targets=[targets], value=node)
+    assign_node = gast.Assign(
+        targets=[targets],
+        value=node,
+        type_comment=None,
+    )
     return targets, assign_node
 
 

--- a/python/paddle/jit/dy2static/utils.py
+++ b/python/paddle/jit/dy2static/utils.py
@@ -303,6 +303,7 @@ def wrap_as_closure(tree: gast.AST, closure_vars: list[str]) -> gast.AST:
                 )
             ],
             value=value,
+            type_comment=None,
         )
 
     def create_wrppper_fn_def_node(name, body) -> gast.FunctionDef:
@@ -321,6 +322,7 @@ def wrap_as_closure(tree: gast.AST, closure_vars: list[str]) -> gast.AST:
             decorator_list=[],
             returns=None,
             type_comment=None,
+            type_params=[],
         )
 
     if not isinstance(tree, gast.Module):

--- a/python/paddle/utils/gast/ast3.py
+++ b/python/paddle/utils/gast/ast3.py
@@ -36,15 +36,36 @@ import sys
 
 
 class Ast3ToGAst(AstToGAst):
+    if sys.version_info.minor < 10:
+
+        def visit_alias(self, node):
+            new_node = gast.alias(
+                self._visit(node.name),
+                self._visit(node.asname),
+            )
+            new_node.lineno = new_node.col_offset = None
+            new_node.end_lineno = new_node.end_col_offset = None
+            return new_node
+
     if sys.version_info.minor < 9:
 
         def visit_ExtSlice(self, node):
             new_node = gast.Tuple(self._visit(node.dims), gast.Load())
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_Index(self, node):
             return self._visit(node.value)
+
+        def visit_Assign(self, node):
+            new_node = gast.Assign(
+                self._visit(node.targets),
+                self._visit(node.value),
+                None,  # type_comment
+            )
+
+            gast.copy_location(new_node, node)
+            new_node.end_lineno = new_node.end_col_offset = None
+            return new_node
 
     if sys.version_info.minor < 8:
 
@@ -57,8 +78,7 @@ class Ast3ToGAst(AstToGAst):
                 node.n,
                 None,
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_Ellipsis(self, node):
             new_node = gast.Constant(
@@ -74,16 +94,14 @@ class Ast3ToGAst(AstToGAst):
                 node.s,
                 None,
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_Bytes(self, node):
             new_node = gast.Constant(
                 node.s,
                 None,
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_FunctionDef(self, node):
             new_node = gast.FunctionDef(
@@ -93,9 +111,9 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.decorator_list),
                 self._visit(node.returns),
                 None,  # type_comment
+                [],  # type_params
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_AsyncFunctionDef(self, node):
             new_node = gast.AsyncFunctionDef(
@@ -105,9 +123,9 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.decorator_list),
                 self._visit(node.returns),
                 None,  # type_comment
+                [],  # type_params
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_For(self, node):
             new_node = gast.For(
@@ -117,8 +135,7 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.orelse),
                 None,  # type_comment
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_AsyncFor(self, node):
             new_node = gast.AsyncFor(
@@ -128,8 +145,7 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.orelse),
                 None,  # type_comment
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_With(self, node):
             new_node = gast.With(
@@ -137,8 +153,7 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.body),
                 None,  # type_comment
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_AsyncWith(self, node):
             new_node = gast.AsyncWith(
@@ -146,8 +161,7 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.body),
                 None,  # type_comment
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_Call(self, node):
             if sys.version_info.minor < 5:
@@ -172,8 +186,7 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.args) + starred,
                 self._visit(node.keywords) + kwargs,
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_NameConstant(self, node):
             if node.value is None:
@@ -182,8 +195,7 @@ class Ast3ToGAst(AstToGAst):
                 new_node = gast.Constant(True, None)
             elif node.value is False:
                 new_node = gast.Constant(False, None)
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
         def visit_arguments(self, node):
             new_node = gast.arguments(
@@ -195,33 +207,30 @@ class Ast3ToGAst(AstToGAst):
                 self._visit(node.kwarg),
                 self._visit(node.defaults),
             )
-            gast.copy_location(new_node, node)
-            return new_node
+            return gast.copy_location(new_node, node)
 
     def visit_Name(self, node):
         new_node = gast.Name(
-            self._visit(node.id),
+            node.id,  # micro-optimization here, don't call self._visit
             self._visit(node.ctx),
             None,
             None,
         )
-        ast.copy_location(new_node, node)
-        return new_node
+        return ast.copy_location(new_node, node)
 
     def visit_arg(self, node):
         if sys.version_info.minor < 8:
-            extra_args = [None]
+            extra_arg = None
         else:
-            extra_args = [self._visit(node.type_comment)]
+            extra_arg = self._visit(node.type_comment)
 
         new_node = gast.Name(
-            self._visit(node.arg),
+            node.arg,  # micro-optimization here, don't call self._visit
             gast.Param(),
             self._visit(node.annotation),
-            *extra_args,  # type_comment
+            extra_arg,  # type_comment
         )
-        ast.copy_location(new_node, node)
-        return new_node
+        return ast.copy_location(new_node, node)
 
     def visit_ExceptHandler(self, node):
         if node.name:
@@ -230,8 +239,7 @@ class Ast3ToGAst(AstToGAst):
                 gast.Name(node.name, gast.Store(), None, None),
                 self._visit(node.body),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
         else:
             return self.generic_visit(node)
 
@@ -246,8 +254,42 @@ class Ast3ToGAst(AstToGAst):
             )
             return ast.copy_location(new_node, node)
 
+    if 8 <= sys.version_info.minor < 12:
+
+        def visit_FunctionDef(self, node):
+            new_node = gast.FunctionDef(
+                self._visit(node.name),
+                self._visit(node.args),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                self._visit(node.returns),
+                self._visit(node.type_comment),
+                [],  # type_params
+            )
+            return gast.copy_location(new_node, node)
+
+        def visit_AsyncFunctionDef(self, node):
+            new_node = gast.AsyncFunctionDef(
+                self._visit(node.name),
+                self._visit(node.args),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                self._visit(node.returns),
+                self._visit(node.type_comment),
+                [],  # type_params
+            )
+            return gast.copy_location(new_node, node)
+
 
 class GAstToAst3(GAstToAst):
+    if sys.version_info.minor < 10:
+
+        def visit_alias(self, node):
+            new_node = ast.alias(
+                self._visit(node.name), self._visit(node.asname)
+            )
+            return new_node
+
     if sys.version_info.minor < 9:
 
         def visit_Subscript(self, node):
@@ -275,8 +317,15 @@ class GAstToAst3(GAstToAst):
                 new_slice,
                 self._visit(node.ctx),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
+
+    def visit_Assign(self, node):
+        new_node = ast.Assign(
+            self._visit(node.targets),
+            self._visit(node.value),
+        )
+
+        return ast.copy_location(new_node, node)
 
     if sys.version_info.minor < 8:
 
@@ -297,8 +346,7 @@ class GAstToAst3(GAstToAst):
                 new_node = ast.Str(node.value)
             else:
                 new_node = ast.Bytes(node.value)
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
     def _make_arg(self, node):
         if node is None:
@@ -319,8 +367,7 @@ class GAstToAst3(GAstToAst):
             self._visit(node.id),
             self._visit(node.ctx),
         )
-        ast.copy_location(new_node, node)
-        return new_node
+        return ast.copy_location(new_node, node)
 
     def visit_ExceptHandler(self, node):
         if node.name:
@@ -355,8 +402,7 @@ class GAstToAst3(GAstToAst):
                 self._visit(starargs),
                 self._visit(kwargs),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_ClassDef(self, node):
             self.generic_visit(node)
@@ -381,8 +427,7 @@ class GAstToAst3(GAstToAst):
                 self._visit(node.decorator_list),
                 self._visit(node.returns),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_AsyncFunctionDef(self, node):
             new_node = ast.AsyncFunctionDef(
@@ -392,8 +437,7 @@ class GAstToAst3(GAstToAst):
                 self._visit(node.decorator_list),
                 self._visit(node.returns),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_For(self, node):
             new_node = ast.For(
@@ -402,8 +446,7 @@ class GAstToAst3(GAstToAst):
                 self._visit(node.body),
                 self._visit(node.orelse),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_AsyncFor(self, node):
             new_node = ast.AsyncFor(
@@ -413,24 +456,21 @@ class GAstToAst3(GAstToAst):
                 self._visit(node.orelse),
                 None,  # type_comment
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_With(self, node):
             new_node = ast.With(
                 self._visit(node.items),
                 self._visit(node.body),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_AsyncWith(self, node):
             new_node = ast.AsyncWith(
                 self._visit(node.items),
                 self._visit(node.body),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
 
         def visit_Call(self, node):
             new_node = ast.Call(
@@ -438,8 +478,43 @@ class GAstToAst3(GAstToAst):
                 self._visit(node.args),
                 self._visit(node.keywords),
             )
-            ast.copy_location(new_node, node)
-            return new_node
+            return ast.copy_location(new_node, node)
+
+    if 5 <= sys.version_info.minor < 12:
+
+        def visit_ClassDef(self, node):
+            new_node = ast.ClassDef(
+                self._visit(node.name),
+                self._visit(node.bases),
+                self._visit(node.keywords),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+            )
+            return ast.copy_location(new_node, node)
+
+    if 8 <= sys.version_info.minor < 12:
+
+        def visit_FunctionDef(self, node):
+            new_node = ast.FunctionDef(
+                self._visit(node.name),
+                self._visit(node.args),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                self._visit(node.returns),
+                self._visit(node.type_comment),
+            )
+            return ast.copy_location(new_node, node)
+
+        def visit_AsyncFunctionDef(self, node):
+            new_node = ast.AsyncFunctionDef(
+                self._visit(node.name),
+                self._visit(node.args),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                self._visit(node.returns),
+                self._visit(node.type_comment),
+            )
+            return ast.copy_location(new_node, node)
 
     def visit_arguments(self, node):
         extra_args = [

--- a/python/paddle/utils/gast/ast3.py
+++ b/python/paddle/utils/gast/ast3.py
@@ -103,6 +103,17 @@ class Ast3ToGAst(AstToGAst):
             )
             return gast.copy_location(new_node, node)
 
+        def visit_ClassDef(self, node):
+            new_node = gast.ClassDef(
+                self._visit(node.name),
+                self._visit(node.bases),
+                self._visit(node.keywords),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                [],  # type_params
+            )
+            return gast.copy_location(new_node, node)
+
         def visit_FunctionDef(self, node):
             new_node = gast.FunctionDef(
                 self._visit(node.name),
@@ -255,6 +266,17 @@ class Ast3ToGAst(AstToGAst):
             return ast.copy_location(new_node, node)
 
     if 8 <= sys.version_info.minor < 12:
+
+        def visit_ClassDef(self, node):
+            new_node = gast.ClassDef(
+                self._visit(node.name),
+                self._visit(node.bases),
+                self._visit(node.keywords),
+                self._visit(node.body),
+                self._visit(node.decorator_list),
+                [],  # type_params
+            )
+            return gast.copy_location(new_node, node)
 
         def visit_FunctionDef(self, node):
             new_node = gast.FunctionDef(

--- a/python/paddle/utils/gast/astn.py
+++ b/python/paddle/utils/gast/astn.py
@@ -34,26 +34,33 @@ from . import gast
 
 
 def _generate_translators(to):
+
     class Translator(ast.NodeTransformer):
+
         def _visit(self, node):
-            if isinstance(node, list):
-                return [self._visit(n) for n in node]
-            elif isinstance(node, ast.AST):
+            if isinstance(node, ast.AST):
                 return self.visit(node)
+            elif isinstance(node, list):
+                return [self._visit(n) for n in node]
             else:
                 return node
 
         def generic_visit(self, node):
             cls = type(node).__name__
-            # handle nodes that are not part of the AST
-            if not hasattr(to, cls):
+            try:
+                new_node = getattr(to, cls)()
+            except AttributeError:
+                # handle nodes that are not part of the AST
                 return
-            new_node = getattr(to, cls)()
+
             for field in node._fields:
                 setattr(new_node, field, self._visit(getattr(node, field)))
-            for attr in getattr(node, '_attributes'):
-                if hasattr(node, attr):
+
+            for attr in node._attributes:
+                try:
                     setattr(new_node, attr, getattr(node, attr))
+                except AttributeError:
+                    pass
             return new_node
 
     return Translator

--- a/python/paddle/utils/gast/gast.py
+++ b/python/paddle/utils/gast/gast.py
@@ -43,29 +43,53 @@ except ImportError:
         pass
 
 
+try:
+    from ast import pattern
+except ImportError:
+
+    class pattern(AST):
+        pass
+
+
+try:
+    from ast import type_param
+except ImportError:
+
+    class type_param(AST):
+        pass
+
+
 def _make_node(Name, Fields, Attributes, Bases):
+
+    # This constructor is used a lot during conversion from ast to gast,
+    # then as the primary way to build ast nodes. So we tried to optimized it
+    # for speed and not for readability.
     def create_node(self, *args, **kwargs):
-        nbparam = len(args) + len(kwargs)
-        assert nbparam in (
-            0,
-            len(Fields),
-        ), "Bad argument number for {}: {}, expecting {}".format(
-            Name, nbparam, len(Fields)
-        )
-        self._fields = Fields
-        self._attributes = Attributes
-        for argname, argval in zip(self._fields, args):
-            setattr(self, argname, argval)
-        for argname, argval in kwargs.items():
-            assert (
-                argname in Fields
-            ), "Invalid Keyword argument for {}: {}".format(Name, argname)
-            setattr(self, argname, argval)
+        if len(args) > len(Fields):
+            raise TypeError(
+                "{} constructor takes at most {} positional arguments".format(
+                    Name, len(Fields)
+                )
+            )
+
+        # it's faster to iterate rather than zipping or enumerate
+        for i in range(len(args)):
+            setattr(self, Fields[i], args[i])
+        if kwargs:  # cold branch
+            self.__dict__.update(kwargs)
 
     setattr(
         _sys.modules[__name__],
         Name,
-        type(Name, Bases, {'__init__': create_node}),
+        type(
+            Name,
+            Bases,
+            {
+                '__init__': create_node,
+                '_fields': Fields,
+                '_attributes': Attributes,
+            },
+        ),
     )
 
 
@@ -87,6 +111,7 @@ _nodes = (
                 'decorator_list',
                 'returns',
                 'type_comment',
+                'type_params',
             ),
             (
                 'lineno',
@@ -107,6 +132,7 @@ _nodes = (
                 'decorator_list',
                 'returns',
                 'type_comment',
+                'type_params',
             ),
             (
                 'lineno',
@@ -126,6 +152,7 @@ _nodes = (
                 'keywords',
                 'body',
                 'decorator_list',
+                'type_params',
             ),
             (
                 'lineno',
@@ -165,10 +192,20 @@ _nodes = (
     (
         'Assign',
         (
+            ('targets', 'value', 'type_comment'),
             (
-                'targets',
-                'value',
+                'lineno',
+                'col_offset',
+                'end_lineno',
+                'end_col_offset',
             ),
+            (stmt,),
+        ),
+    ),
+    (
+        'TypeAlias',
+        (
+            ('name', 'type_params', 'value'),
             (
                 'lineno',
                 'col_offset',
@@ -317,6 +354,19 @@ _nodes = (
         ),
     ),
     (
+        'Match',
+        (
+            ('subject', 'cases'),
+            (
+                'lineno',
+                'col_offset',
+                'end_lineno',
+                'end_col_offset',
+            ),
+            (stmt,),
+        ),
+    ),
+    (
         'Raise',
         (
             (
@@ -334,6 +384,24 @@ _nodes = (
     ),
     (
         'Try',
+        (
+            (
+                'body',
+                'handlers',
+                'orelse',
+                'finalbody',
+            ),
+            (
+                'lineno',
+                'col_offset',
+                'end_lineno',
+                'end_col_offset',
+            ),
+            (stmt,),
+        ),
+    ),
+    (
+        'TryStar',
         (
             (
                 'body',
@@ -498,6 +566,22 @@ _nodes = (
             (
                 'op',
                 'values',
+            ),
+            (
+                'lineno',
+                'col_offset',
+                'end_lineno',
+                'end_col_offset',
+            ),
+            (expr,),
+        ),
+    ),
+    (
+        'NamedExpr',
+        (
+            (
+                'target',
+                'value',
             ),
             (
                 'lineno',
@@ -1024,29 +1108,133 @@ _nodes = (
         ),
     ),
     # alias
-    ('alias', (('name', 'asname'), (), (AST,))),
+    (
+        'alias',
+        (
+            ('name', 'asname'),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (AST,),
+        ),
+    ),
     # withitem
     ('withitem', (('context_expr', 'optional_vars'), (), (AST,))),
+    # match_case
+    ('match_case', (('pattern', 'guard', 'body'), (), (AST,))),
+    # pattern
+    (
+        'MatchValue',
+        (
+            ('value',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchSingleton',
+        (
+            ('value',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchSequence',
+        (
+            ('patterns',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchMapping',
+        (
+            ('keys', 'patterns', 'rest'),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchClass',
+        (
+            ('cls', 'patterns', 'kwd_attrs', 'kwd_patterns'),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchStar',
+        (
+            ('name',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchAs',
+        (
+            ('pattern', 'name'),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
+    (
+        'MatchOr',
+        (
+            ('patterns',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (pattern,),
+        ),
+    ),
     # type_ignore
     ('type_ignore', ((), ('lineno', 'tag'), (TypeIgnore,))),
+    # type_param
+    (
+        'TypeVar',
+        (
+            (
+                'name',
+                'bound',
+            ),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (type_param,),
+        ),
+    ),
+    (
+        'ParamSpec',
+        (
+            ('name',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (type_param,),
+        ),
+    ),
+    (
+        'TypeVarTuple',
+        (
+            ('name',),
+            ('lineno', 'col_offset', 'end_lineno', 'end_col_offset'),
+            (type_param,),
+        ),
+    ),
 )
+
 
 for name, descr in _nodes:
     _make_node(name, *descr)
 
-py_version = _sys.version_info.major
-if py_version != 3:
-    raise RuntimeError(
-        'Required Python version >= 3, but received Python version == {}'.format(
-            py_version
-        )
-    )
-
-from .ast3 import ast_to_gast, gast_to_ast
+if _sys.version_info.major == 2:
+    from .ast2 import ast_to_gast, gast_to_ast
+if _sys.version_info.major == 3:
+    from .ast3 import ast_to_gast, gast_to_ast
 
 
 def parse(*args, **kwargs):
     return ast_to_gast(_ast.parse(*args, **kwargs))
+
+
+def unparse(gast_obj):
+    from .unparser import unparse
+
+    return unparse(gast_obj)
 
 
 def literal_eval(node_or_string):
@@ -1139,3 +1327,111 @@ def increment_lineno(node, n=1):
         if 'end_lineno' in child._attributes:
             child.end_lineno = (getattr(child, 'end_lineno', 0) or 0) + n
     return node
+
+
+if _sys.version_info.major == 3 and _sys.version_info.minor >= 13:
+    dump = _ast.dump
+else:
+    # Code import from Lib/ast.py
+    #
+    # minor changes: getattr(x, y, ...) is None => getattr(x, y, 42) is None
+    #
+    def dump(
+        node,
+        annotate_fields=True,
+        include_attributes=False,
+        # *,  # removed for compatibility with python2 :-/
+        indent=None,
+        show_empty=False,
+    ):
+        """
+        Return a formatted dump of the tree in node.  This is mainly useful for
+        debugging purposes.  If annotate_fields is true (by default),
+        the returned string will show the names and the values for fields.
+        If annotate_fields is false, the result string will be more compact by
+        omitting unambiguous field names.  Attributes such as line
+        numbers and column offsets are not dumped by default.  If this is wanted,
+        include_attributes can be set to true.  If indent is a non-negative
+        integer or string, then the tree will be pretty-printed with that indent
+        level. None (the default) selects the single line representation.
+        If show_empty is False, then empty lists and fields that are None
+        will be omitted from the output for better readability.
+        """
+
+        def _format(node, level=0):
+            if indent is not None:
+                level += 1
+                prefix = '\n' + indent * level
+                sep = ',\n' + indent * level
+            else:
+                prefix = ''
+                sep = ', '
+            if isinstance(node, AST):
+                cls = type(node)
+                args = []
+                args_buffer = []
+                allsimple = True
+                keywords = annotate_fields
+                for name in node._fields:
+                    try:
+                        value = getattr(node, name)
+                    except AttributeError:
+                        keywords = True
+                        continue
+                    if value is None and getattr(cls, name, 42) is None:
+                        keywords = True
+                        continue
+                    if (
+                        not show_empty
+                        and (value is None or value == [])
+                        # Special cases:
+                        # `Constant(value=None)` and `MatchSingleton(value=None)`
+                        and not isinstance(node, (Constant, MatchSingleton))
+                    ):
+                        args_buffer.append(repr(value))
+                        continue
+                    elif not keywords:
+                        args.extend(args_buffer)
+                        args_buffer = []
+                    value, simple = _format(value, level)
+                    allsimple = allsimple and simple
+                    if keywords:
+                        args.append('%s=%s' % (name, value))
+                    else:
+                        args.append(value)
+                if include_attributes and node._attributes:
+                    for name in node._attributes:
+                        try:
+                            value = getattr(node, name)
+                        except AttributeError:
+                            continue
+                        if value is None and getattr(cls, name, 42) is None:
+                            continue
+                        value, simple = _format(value, level)
+                        allsimple = allsimple and simple
+                        args.append('%s=%s' % (name, value))
+                if allsimple and len(args) <= 3:
+                    return (
+                        '%s(%s)' % (node.__class__.__name__, ', '.join(args)),
+                        not args,
+                    )
+                return (
+                    '%s(%s%s)'
+                    % (node.__class__.__name__, prefix, sep.join(args)),
+                    False,
+                )
+            elif isinstance(node, list):
+                if not node:
+                    return '[]', True
+                return (
+                    '[%s%s]'
+                    % (prefix, sep.join(_format(x, level)[0] for x in node)),
+                    False,
+                )
+            return repr(node), True
+
+        if not isinstance(node, AST):
+            raise TypeError('expected AST, got %r' % node.__class__.__name__)
+        if indent is not None and not isinstance(indent, str):
+            indent = ' ' * indent
+        return _format(node)[0]

--- a/test/legacy_test/test_gast_with_compatibility.py
+++ b/test/legacy_test/test_gast_with_compatibility.py
@@ -13,7 +13,6 @@
 # limitations under the License.
 
 import ast
-import sys
 import textwrap
 import unittest
 
@@ -142,60 +141,47 @@ class TestPythonCompatibility(unittest.TestCase):
         """
         self._check_compatibility(source, target)
 
-    # The 0.3.3 version of gast has a bug in python3.8 that
-    # would cause the following tests to fail. But this
-    # problem doesn't affect the use of Paddle's related
-    # functions, therefore, the following tests would be
-    # disable in python3.8.
-    #
-    # This problem had been fixed and updated to version
-    # 0.4.1 of gast.
-    #
-    # More information please refer to:
-    # https://github.com/serge-sans-paille/gast/issues/49
-    if sys.version_info >= (3, 9):
+    def test_with(self):
+        """
+        The fileds `context_expr/optional_vars` of `ast.With` in PY2
+        is moved into `ast.With.items.withitem` in PY3.
+        """
+        source = """
+        with guard():
+            a = 1
+        """
+        target = """
+        with guard_new():
+            a = 1
+        """
+        self._check_compatibility(source, target)
 
-        def test_with(self):
-            """
-            The fileds `context_expr/optional_vars` of `ast.With` in PY2
-            is moved into `ast.With.items.withitem` in PY3.
-            """
-            source = """
-            with guard():
-                a = 1
-            """
-            target = """
-            with guard_new():
-                a = 1
-            """
-            self._check_compatibility(source, target)
+    def test_subscript_Index(self):
+        source = """
+            x = y()[10]
+        """
+        target = """
+            x = y()[20]
+        """
+        self._check_compatibility(source, target)
 
-        def test_subscript_Index(self):
-            source = """
-                x = y()[10]
-            """
-            target = """
-                x = y()[20]
-            """
-            self._check_compatibility(source, target)
+    def test_subscript_Slice(self):
+        source = """
+            x = y()[10:20]
+        """
+        target = """
+            x = y()[20:40]
+        """
+        self._check_compatibility(source, target)
 
-        def test_subscript_Slice(self):
-            source = """
-                x = y()[10:20]
-            """
-            target = """
-                x = y()[20:40]
-            """
-            self._check_compatibility(source, target)
-
-        def test_call(self):
-            source = """
-                y = foo(*arg)
-            """
-            target = """
-                y = foo(*arg_new)
-            """
-            self._check_compatibility(source, target)
+    def test_call(self):
+        source = """
+            y = foo(*arg)
+        """
+        target = """
+            y = foo(*arg_new)
+        """
+        self._check_compatibility(source, target)
 
 
 if __name__ == '__main__':

